rust/lance-linalg/src/kmeans.rs

```diff
@@ -446,37 +447,44 @@ impl KMeans {
         let n = data.len() / self.dimension;
         let metric_type = self.metric_type;
         const CHUNK_SIZE: usize = 1024;
+
+        // Normalized centroids for fast cosine.
+        let norm_centroids = if matches!(metric_type, MetricType::Cosine) {
+            Arc::new(Some(
+                self.centroids
+                    .values()
+                    .chunks_exact(dimension)
+                    .map(|centroid| centroid.norm_l2())
+                    .collect::<Vec<f32>>(),
+            ))
+        } else {
+            Arc::new(None)
+        };
+
         let cluster_with_distances = stream::iter((0..n).step_by(CHUNK_SIZE))
             // make tiles of input data to split between threads.
-            .zip(repeat_with(|| (data.clone(), self.centroids.clone())))
-            .map(|(start_idx, (data, centroids))| async move {
+            .zip(repeat_with(|| {
+                (data.clone(), self.centroids.clone(), norm_centroids.clone())
+            }))
+            .map(|(start_idx, (data, centroids, norms))| async move {
                 let data = tokio::task::spawn_blocking(move || {
                     let array = data.values();
                     let centroids_array = centroids.values();
 
                     (start_idx..min(start_idx + CHUNK_SIZE, n))
                         .map(|idx| {
                             let vector = &array[idx * dimension..(idx + 1) * dimension];
-                            let mut min = std::f32::MAX;
-                            let mut min_idx = 0;
-                            for (idx, other) in centroids_array.chunks_exact(dimension).enumerate()
-                            {
-                                // We've found about 40% performance improvement by using static dispatch instead
-                                // of dynamic dispatch.
-                                //
-                                // NOTE: Please make sure run benchmark when changing the following code.
-                                // `RUSTFLAGS="-C target-cpu=native" cargo bench --bench ivf_pq`
```

Comment:
should we keep this note here?

+ have we tested branching cost? i.e. if we have a generic `KMeans<DistanceType=...>` vs match statement?

Reply:
+1 on this. For critical pieces of code, I bet it's worth not having the `match` in the loop.

Reply:
> have we tested branching cost? i.e. if we have a generic KMeans<DistanceType=...> vs match statement?

Can do in follow up PRs.

This one meant to make #1312 smaller / reviewable.

Reply:
added the comment back